crates/binstalk-downloader/src/remote/resolver.rs

```diff
@@ -23,42 +26,75 @@ impl Resolve for TrustDnsResolver {
-            .iter()
-            .filter_map(|adapter| {
-                if adapter.adapter_name() == current_interface.name {
-                    Some(adapter.dns_servers())
-                } else {
-                    None
-                }
-            })
-            .flatten()
-            .for_each(|addr| {
-                let socket_addr = SocketAddr::new(*addr, 53);
-                for protocol in [Protocol::Udp, Protocol::Tcp] {
-                    config.add_name_server(NameServerConfig {
-                        socket_addr,
-                        protocol,
-                        tls_dns_name: None,
-                        trust_negative_responses: false,
-                        #[cfg(feature = "rustls")]
-                        tls_config: None,
-                        bind_addr: None,
-                    })
-                }
-            });
+        get_dns_servers()?.into_iter().for_each(|addr| {
+            trace!("Adding DNS server: {}", addr);
+            let socket_addr = SocketAddr::new(addr, 53);
+            for protocol in [Protocol::Udp, Protocol::Tcp] {
+                config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol,
+                    tls_dns_name: None,
+                    trust_negative_responses: false,
+                    #[cfg(feature = "rustls")]
+                    tls_config: None,
+                    bind_addr: None,
+                })
+            }
+        });
 
+        debug!("Resolver configuration complete");
         Ok(TokioAsyncResolver::tokio(config, opts))
     }
 }
+
+#[cfg(windows)]
+#[instrument(level = "trace")]
+fn get_dns_servers() -> Result<Vec<std::net::IpAddr>, BoxError> {
+    debug!("Retrieving local IP address");
+    let local_ip = match default_net::interface::get_local_ipaddr() {
+        Some(ip) => ip,
+        None => return Err("Local IP address not found".into()),
+    };
+    debug!("Local IP address: {local_ip}");
+    debug!("Retrieving network adapters");
+    let adapters = ipconfig::get_adapters()?;
+    debug!("Found {} network adapters", adapters.len());
+    debug!("Searching for adapter with IP address {local_ip}");
+    let adapter = find_adapter(local_ip, adapters)?;
+    debug!(
+        "Using adapter {} with {} DNS servers",
+        adapter.friendly_name(),
+        adapter.dns_servers().len()
+    );
+    Ok(adapter.dns_servers().to_vec())
+}
+
+#[cfg(windows)]
+fn find_adapter(
+    local_ip: std::net::IpAddr,
+    adapters: Vec<ipconfig::Adapter>,
+) -> Result<ipconfig::Adapter, BoxError> {
+    let mut found: Option<ipconfig::Adapter> = None;
+    for adapter in adapters {
+        if adapter.ip_addresses().contains(&local_ip) {
+            debug!("Found adapter candidate: {}", adapter.friendly_name());
+            if found.is_none() {
+                found = Some(adapter);
+            }
+        }
+    }
+    found.ok_or_else(|| "Adapter not found".into())
```

Comment:
Using [`Iterator::find`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) would simplify this a lot:

```suggestion
    Ok(
        adapters.into_iter()
            .find(|adapter| adapter.ip_addresses().contains(&local_ip))
            .ok_or("Adapter not found")?
    )
```

Reply:
I did this intentionally (since this wasn't originally meant to be a fix, but to find the problem) so I could log all of the adapters that fulfill the criteria. Since this is no longer necessary, `Iterator::find` can be used now like you said.